File: iosApp/iosApp/Modules/KeyStore/KeyStoreViewModel.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation

struct KeyStoreViewModel {
    let isEmpty: Bool
    let state: State
    let items: [KeyStoreItem]
    let selectedIdxs: [Int]
    let buttons: ButtonSheetViewModel
    let targetView: TransitionTargetView
}

extension KeyStoreViewModel {

    enum State {
        case loading
        case loaded
        case error(error: KeyStoreViewModel.Error)
    }
}

extension KeyStoreViewModel {

    struct KeyStoreItem {
        
        let title: String
        let address: String?
    }
}

extension KeyStoreViewModel {

    struct Error {
        let title: String
        let body: String
        let actions: [String]
    }
}

extension KeyStoreViewModel {

    enum TransitionTargetView {
        case keyStoreItemAt(idx: Int)
        case buttonAt(idx: Int)
        case none
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/keyStore/KeyStoreViewModel.kt
package com.sonsofcrypto.web3walletcore.modules.keyStore

import com.sonsofcrypto.web3walletcore.common.viewModels.ErrorViewModel

data class KeyStoreViewModel(
    val isEmpty: Boolean,
    val state: State,
    val items: List<Item>,
    val selectedIdxs: List<Int>,
    val buttons: ButtonSheetViewModel,
    val targetView: TransitionTargetView,
) {
    sealed class State {
        object Loading: State()
        object Loaded: State()
        data class Error(val error: ErrorViewModel): State()
    }

    data class Item(
        val title: String,
        val address: String?,
    )

    data class ButtonSheetViewModel(
        val buttons: List<Button>,
        val mode: SheetMode,
    ) {
        data class Button(
            val title: String,
            val type: Type,
        ) {
            enum class Type {
                NEW_MNEMONIC,
                IMPORT_MNEMONIC,
                MORE_OPTION,
                CONNECT_HARDWARE_WALLET,
                IMPORT_PRIVATE_KEY,
                CREATE_MULTI_SIG,
            }
        }

        enum class SheetMode { HIDDEN, COMPACT, EXPANDED }
    }

    sealed class TransitionTargetView {
        data class KeyStoreItemAt(val idx: Int): TransitionTargetView()
        data class ButtonAt(val idx: Int): TransitionTargetView()
        object None: TransitionTargetView()
    }
}
